torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -347,6 +355,260 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+    case at::kBool: memcpy_cpu<bool> (tensor_ptr, ndarray_ptr, num_items, itemsize, isaligned); return (void*)(((bool *)tensor_ptr) + num_items);
+    case at::kBFloat16: memcpy_cpu<at::BFloat16> (tensor_ptr, ndarray_ptr, num_items, itemsize, isaligned); return (void*)(((at::BFloat16 *)tensor_ptr) + num_items);
+    default: throw std::runtime_error("invalid type");
+  }
+  return NULL;
+}
+
+
+Tensor reverse_strides(Tensor & tensor) {
+  /*
+  Utility function for reversing the strides of a tensor excluding the outermost dimension.
+  So, a tensor with strides [3200, 1200, 720, 80] will return a tensor with strides reversed: [3200, 80, 720, 1200]
+  This is used when copying F-contiguous numpy arrays to the output tensor in tensor_from_ndarray_batch().
+  */
+  auto strides = tensor.strides();
+  auto size = strides.size();
+  std::vector<int64_t> new_strides(size);
+  new_strides[0] = strides[0];
+  for (int64_t i = 0; i < size - 1; ++i) {
+    new_strides[i+1] = strides[size-i-1];
+  }
+  return at::native::set_storage_cpu_(tensor, tensor.storage(), tensor.storage_offset(), tensor.sizes(), new_strides);
+}
+
+
+at::Tensor tensor_from_ndarray_batch(std::vector<PyArrayObject*> & array_ptr_storage, std::vector<int64_t> & sizes, ScalarType scalarType, bool pin_memory) {
+  /*
+  Returns a newly generated tensor by first allocating an empty one and then filling up with numpy array elements using raw memory copy. This function is kept
+  simple enough by delegating the responsibility of ensuring contiguous, aligned arrays whenever a non-contiguous, unaligned arrays are given.
+  See extract_ndarrays and _extract_ndarrays for more information on extracting all the numpy array pointers. This is used in internal_new_from_data from
+  torch/csrc/utils/tensor_new.cpp
+  */
+  size_t C_contiguous_count = 0;
+  bool C_contiguous_majority = true;
+
+  for (auto array:array_ptr_storage) {
+    if (PyArray_IS_C_CONTIGUOUS(array)) {
+      C_contiguous_count += 1;
+    }
+  }
+
+  if (2*C_contiguous_count < array_ptr_storage.size()) {
+    C_contiguous_majority = false;
+  }
+
+  auto tensor = at::empty(sizes, at::initialTensorOptions().dtype(scalarType).pinned_memory(pin_memory));
+  int np_type = aten_to_numpy_dtype(scalarType);
+  auto tensor_ptr = tensor.data_ptr();
+
+  for (auto array:array_ptr_storage) {
+    auto flags = PyArray_FLAGS(array);
+    if (C_contiguous_majority) {
+        auto array_tmp = (PyArrayObject*) PyArray_FromArray(array, PyArray_DescrFromType(np_type), NPY_ARRAY_CARRAY);        
+        if (!array_tmp) {
+          throw python_error();
+        }
+        tensor_ptr = store_ndarray(tensor_ptr, PyArray_DATA(array_tmp), scalarType, PyArray_SIZE(array_tmp), tensor.dtype().itemsize(), true);
+        if (array_tmp != array) { 
+          Py_DECREF((PyObject*) array_tmp);
+        }
+    } else {
+        auto array_tmp = (PyArrayObject*) PyArray_FromArray(array, PyArray_DescrFromType(np_type), NPY_ARRAY_FARRAY);
```

Comment:
all discontiguous cases (not just f-contiguous) go here, which leads to some unexpected strides in the result:
```
In [16]: a=np.random.rand(2,3,4)[:,:,0]                                                                                                                                                                      

In [17]: a.strides                                                                                                                                                                                           
Out[17]: (96, 32)

In [18]: lists=[a for _ in range(10)]                                                                                                                                                                        

In [19]: b=torch.tensor(lists)                                                                                                                                                                               

In [20]: b.stride()                                                                                                                                                                                          
Out[20]: (6, 1, 3)
```

Reply:
"all discontiguous cases (not just f-contiguous) go here" : That part was intentional. There was a flaw in the way the strides were assigned whenever any non C-contiguous array were used. It somehow eluded the test cases that I wrote. I have changed the way the strides are computed whenever C-contiguous arrays are in minority. For the majority case, I don't have to change the strides since any non C-contiguous arrays are mapped into a C-contiguous array. I have also added extra tests cases including the one that you just wrote above. All tests are passing. Let me know if I should add any other tests.

Reply:
I have limited the optimization to the case that the input list contains numpy arrays of the same shape being the leaves at the same level. The same shape and same leaf level constraints are required for this new stride assigning strategy to work correctly.

Reply:
I'm unsure this `else` branch makes sense. It avoids array copy in `PyArray_FromArray` if the input lists consists of (mostly) f-contiguous arrays, which is likely a rarely encountered niche case. All other discontiguous inputs will first be copied into f-contiguous form, and then sent to store_ndarray. The resulting pytorch tensor will be discontiguous, which, again, likely is not intended outcome. If this `else` branch didn't exist, then those discontiguous cases would be copied to c-contiguous, and result would be contiguous tensor, which is same amount of memory traffic and arguably better output. So, roughly, now:
1) c-contiguous inputs -> no copy by PyArray_FromArray, contiguous output (good!)
2) f-contiguous inputs -> no copy by PyArray_FromArray in the `else` branch, discontiguous output (good, but niche)
3) discontiguous inputs -> copy by PyArray_FromArray in the else branch, discontiguous output (unexpected). 
If you remove `else` branch
1) c-contiguous inputs -> no copy by PyArray_FromArray, contiguous output (good!)
2) f-contiguous inputs -> copy by PyArray_FromArray, contiguous output (worse than above, but code is also simpler)
3) discontiguous inputs -> copy by PyArray_FromArray, contiguous output (same memory traffic as above, better output format). 
Your thoughts?
Also, can you add test cases for discontiguous inputs if they are not present already?

Reply:
Let me state the logic here and the rationale behind it. The idea is that the input can consist of c-contiguous, f-contiguous and discontiguous arrays and that we need to either map all these arrays into aligned, c-contiguous / f-contiguous arrays since some of the arrays might not be aligned (considering all cases). The majority voting strategy has been used for obvious reasons to minimize the extra conversion overhead. Two reasons for converting to c-contiguous/f-contiguous arrays : 

1. The ```store_ndarray``` copying mechanism assumes that the arrays are contiguous (c or f).
2. The resulting numpy arrays must all have the same memory layout - either all being c-contiguous or all being f-contiguous. This is required due to way the strides of the tensor are computed. 

Currently, with the newest commit, here is what happens:
1. mostly c-contiguous inputs -> no copy by PyArray_FromArray, c-contiguous output (As you said earlier). Any non c-contiguous input will be made c-contiguous (by copying if necessary).
2. mostly f-contiguous inputs -> no copy by PyArray_FromArray in the else branch, f-contiguous output. Any non f-contiguous (that includes c-contiguous and discontiguous arrays) input will be made f-contiguous (by copying if necessary). While f-contiguous inputs are less common, the most common case is the transpose of a c-contiguous numpy array. 
3. With the latest commit : discontiguous inputs -> copy by PyArray_FromArray, c-contiguous output.

Now if I remove the ```else``` branch, any f-contiguous array needs to be mapped to a c-contiguous format even if all of them are f-contiguous . All of them might be the transpose of a c-contiguous array. Now that I have gotten the striding strategy right, I am a bit inclined to keep the ```else``` branch.

Test cases for c-contiguous, f-contiguous and discontiguous arrays are present currently.